src/arch/arm/object/vcpu.c

```diff
@@ -378,8 +378,10 @@ exception_t invokeVCPUInjectIRQ(vcpu_t *vcpu, unsigned long index, virq_t virq)
     if (likely(ARCH_NODE_STATE(armHSCurVCPU) == vcpu)) {
         set_gic_vcpu_ctrl_lr(index, virq);
 #ifdef ENABLE_SMP_SUPPORT
-    } else if (vcpu->vcpuTCB->tcbAffinity != getCurrentCPUIndex()) {
-        doRemoteOp3Arg(IpiRemoteCall_VCPUInjectInterrupt, (word_t)vcpu, index, virq.words[0],      vcpu->vcpuTCB->tcbAffinity);
+    } else if (vcpu->vcpuTCB != NULL && vcpu->vcpuTCB->tcbAffinity != getCurrentCPUIndex()) {
```

Comment:
Doesn't this branch only need to be taken if the vcpu is active on another core? If the vcpu is inactive on the other core then the final `else` branch should be sufficient?

Reply:
If you have a multicore guest, then this is the path you need to implement IPIs for the guest.

Reply:
If you're guaranteed that the active VCPU on the other core doesn't change while you're doing that check (I guess that is true since we're inside the lock), then yes, we could add another conjunct to test if it's active on that other core. That would avoid an IPI in those cases.

Reply:
This would delay the vIRQ delivery since we cannot access the physical LR of the host pCPU. The vIRQ would only be injected on the next vCPU entry. The time depends on the number of threads running in the target pCPU and their time slice.

And would also make the removal of big kernel lock even more challenging. Maybe you guys don't care about this.

So far I have been following this protocol, only the host pCPU can change LRs of a vCPU.

I think IPI is important here.

Reply:
> Edit: I think I got your point now, only if it is a active vCPU in the target we send the IPI, otherwise we write in the shadow registers.

Yea, I mean in the case where the ipi handler doesn't access the physical LRs and just writes to memory.

Reply:
Actually, I don't think the code makes much sense for ARM where the VMM task calling `seL4_ARM_VCPU_InjectIRQ()` is a different TCB than the VM itself. I don't think `ARCH_NODE_STATE(armHSCurVCPU) == vcpu` can ever be true except if the VM is injecting IRQs into itself, which is very unlikely.

it seems the whole function can be reduced to just always doing `vcpu->vgic.lr[index] = virq;`? What am I missing?

> This would delay the vIRQ delivery since we cannot access the physical LR of the host pCPU. The vIRQ would only be injected on the next vCPU entry.

I suspect this always happens now, except for cross-core IRQ injection. If the VMM is running, the VM isn't, and it is up to the VMM to schedule the VM so it can receive the IRQ.

Reply:
> it seems the whole function can be reduced to just always doing `vcpu->vgic.lr[index] = virq;`? What am I missing?

When a thread with an associated VCPU switches to a thread without an associated VCPU, the VCPU is disabled, but it's context is left in system registers. So armHSCurVCPU would typically always refer to the last VCPU that was run (unless the object was deleted).

Reply:
`ARCH_NODE_STATE(armHSCurVCPU) == vcpu` can be true when the VMM itself doesn't have a VCPU associated, because VCPU switching is done lazily. In that case `armHSVCPUActive` would be false.

Reply:
I think I mixed up TCB with tcbVCPU when reading `vcpu_switch()`, you're right of course.